compiler/sem/sigmatch.nim

```diff
@@ -2364,68 +2364,116 @@ proc paramTypesMatchAux(m: var TCandidate, f, a: PType,
+      # possibly symbols that can me matched on their position.
+      best = newCandidate(c, candidate.callee)
+      second = newCandidate(c, candidate.callee)
+      # Keep track of the best argument position as well, it will be used
+      # to get back to the selected symbol node.
+      bestArg = -1
+
+    best.calleeSym = candidate.calleeSym
+    second.calleeSym = candidate.calleeSym
+
+    for i in 0 ..< arg.len:
+      # Try all match candidates in the open/closed symbol choice passed as
+      # an argument
+      let matchCandidate = arg[i]
+      if matchCandidate.sym.kind in OverloadableSyms - {skModule}:
+        var next = newCandidate(c, candidate.callee)
+        copyCandidate(next, candidate)
+        next.callee = matchCandidate.typ
+
+        if tfUnresolved in next.callee.flags:
+          continue
+
+        next.calleeSym = matchCandidate.sym
         # XXX this is still all wrong: (T, T) should be 2 generic matches
         # and  (int, int) 2 exact matches, etc. Essentially you cannot call
         # typeRel here and expect things to work!
-        let r = typeRel(z, f, arg[i].typ)
-        incMatches(z, r, 2)
-        if r != isNone:
-          z.state = csMatch
-          case x.state
-          of csEmpty, csNoMatch:
-            x = z
-            best = i
-          of csMatch:
-            let cmp = cmpCandidates(x, z)
-            if cmp < 0:
-              best = i
-              x = z
-            elif cmp == 0:
-              y = z           # z is as good as x
-
-    if x.state == csEmpty:
+        let rel = typeRel(next, formal, matchCandidate.typ)
+        incMatches(next, rel, 2)
+        if rel != isNone:
+          next.state = csMatch
+          case best.state:
+            of csEmpty, csNoMatch:
+              best = next
+              bestArg = i
+
+            of csMatch:
+              let cmp = cmpCandidates(best, next)
+              if cmp < 0:
+                bestArg = i
+                best = next
+
+              elif cmp == 0:
+                second = next           # next is as good as x
+
+    #
+    if best.state == csEmpty:
       result = nil
-    elif y.state == csMatch and cmpCandidates(x, y) == 0:
-      m.c.graph.config.internalAssert(x.state == csMatch, arg.info, "x.state is not csMatch")
+
+    elif second.state == csMatch and
+         # The 'best' and 'second' candidates match equally well - unless
+         # it is a template this means code is ambiguous.
+         cmpCandidates(best, second) == 0:
+
+      candidate.c.graph.config.internalAssert(
+        best.state == csMatch,
+        arg.info,
+        "best.state is not csMatch")
       # ambiguous: more than one symbol fits!
       # See tsymchoice_for_expr as an example. 'f.kind == tyUntyped' should match
       # anyway:
-      if f.kind in {tyUntyped, tyTyped}: result = arg
-      else: result = nil
+      if formal.kind in {tyUntyped, tyTyped}:
+        result = arg
+
+      else:
+        result = nil
```

Comment:
```suggestion
      result =
        if formal.kind in {tyUntyped, tyTyped}:
          arg
        else:
           nil
```

Spacing these small bits out mechanical hurts readability. Broad spaces intimate a significant logical break, which this is not, we're doing one logical thing and that's setting `result`.

Reply:
What do you think about more active usage of the `tern` template for one-line conditionals like this?

`result = tern(formal.kind in {tyUntyped, tyTyped}, arg, nil)`

Reply:
I'd prefer it if it was `ternary` in full or just looked more like the ternary operator that's broadly available in other languages. The latter is unlikely.

With compact code space sensitivity really sucks because space is really inefficient (by the volume of screen real estate required) at expressing logical subdivisions.

This problem could be somewhat alleviated when the parser becomes grammar driven and we can generate code from the grammar including syntax highlighting. Then perhaps we can use that to accentuate the separation.